src/uct/ugni/smsg/ugni_smsg_iface.c

```diff
@@ -164,12 +164,17 @@ UCS_CLASS_DEFINE_DELETE_FUNC(uct_ugni_smsg_iface_t, uct_iface_t);
 static void uct_ugni_smsg_progress(void *arg)
 {
     uct_ugni_smsg_iface_t *iface = (uct_ugni_smsg_iface_t *)arg;
+    uint32_t attempts;
 
     progress_local_cq(iface);
     progress_remote_cq(iface);
 
     /* have a go a processing the pending queue */
-    ucs_arbiter_dispatch(&iface->super.arbiter, 1, uct_ugni_ep_process_pending, NULL);
+    iface->super.send_in_pending = 1;
+    attempts = ucs_min(iface->config.smsg_max_credit, iface->super.send_q_len);
+    ucs_arbiter_dispatch(&iface->super.arbiter, attempts,
+                         uct_ugni_ep_process_pending, NULL);
+    iface->super.send_in_pending = 0;
```

Comment:
@shamisp The problem is that if the remote side progresses and can receive new messages, but the local side has not progressed and has not dispatched jobs from the arbiter than it is possible for a new send to be able to complete, totally bypassing the arbiter and sending the new messages before the queued messages. 

The in progress part of the routine is necessary because if you do not have this than when you attempt to dispatch a job from the arbiter in the send function it will recurse until the OOM killer murders your job because it will always detect a job in the arbiter and it will forever call arbiter dispatch until the OOM reaper has its say.

The trade off is that if you don't do this, and there is any job in the arbiter, than all sends will stall until the local end point has progressed, even if the remote end could receive messages.

This is exactly how this is addressed in the UD transport in IB, as seen in uct_ud_iface_progress_pending_tx called from places like uct_ud_mlx5_ep_am_short